bluesky/traffic/route.py

```diff
@@ -253,15 +270,227 @@ def afteraddwptStack(self, idx, *args):  # args: all arguments of addwpt
         if len(args) > 4:
             arglst[2] = args[4]  # spd
 
-        result = self.addwptStack(idx, *arglst)  # args: all arguments of addwpt
+        result = self.addwptStack(
+            idx, *arglst)  # args: all arguments of addwpt
+
+        return result
+
+    def beforeaddwptStack(self, idx, *args):  # args: all arguments of addwpt
+        """
+            Adds a waypoint before another,
+            using BEFORE command from stack, given as (idx, args).
+            BEFORE acid, wpinroute ADDWPT acid, (wpname/lat,lon),[alt],[spd]
+        """
+        if len(args) < 3:
+            return False, "BEFORE needs more arguments"
+
+        # Change order of arguments
+        arglst = [args[2], None, None, '', args[0]]  # postxt,,,,beforewp
+
+        # Add alt when given
+        if len(args) > 3:
+            arglst[1] = args[3]  # alt
+
+        # Add speed when given
+        if len(args) > 4:
+            arglst[2] = args[4]  # spd
+
+        result = self.addwptStack(
+            idx, *arglst)  # args: all arguments of addwpt
+
+        return result
+
+    def addwpt_data(self, overwrt, wpidx, wpname_, wplat_, wplon_, wptype_,
+                    wpalt_, wpspd_, swflyby_):
+        """
+        Overwrites or inserts information for a waypoint
+        """
+        wplat_ = (wplat_ + 90.) % 180. - 90.
+        wplon_ = (wplon_ + 180.) % 360. - 180.
+
+        if overwrt:
+            self.wpname[wpidx] = wpname_
+            self.wplat[wpidx] = wplat_
+            self.wplon[wpidx] = wplon_
+            self.wpalt[wpidx] = wpalt_
+            self.wpspd[wpidx] = wpspd_
+            self.wptype[wpidx] = wptype_
+            self.wpflyby[wpidx] = swflyby_
+        else:
+            self.wpname.insert(wpidx, wpname_)
+            self.wplat.insert(wpidx, wplat_)
+            self.wplon.insert(wpidx, wplon_)
+            self.wpalt.insert(wpidx, wpalt_)
+            self.wpspd.insert(wpidx, wpspd_)
+            self.wptype.insert(wpidx, wptype_)
+            self.wpflyby.insert(wpidx, swflyby_)
+
+    def addwpt_orig_dest(self, name_, wprtename, wptype, lat, lon, alt, spd):
+        """

```

Comment:
This is essentially the same code as before - mainly it's been just refactored out of `addwpt`.  
As there is fair commonality between adding of origin and destination waypoints, I have combined these actions into a common function, where variables such as `wpidx` are appropriately set based on type (i.e. to 0 for Route.orig and -1 for Route.dest).  

A few things have been tweaked such as incrementing `nwp` only if inserting a wp, not overwriting.